Add Ernie-M Model to huggingface (#21349)
huggingface/transformers: docs/source/en/model_doc/ernie_m.mdx

<!--Copyright 2023 The HuggingFace and Baidu Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# ErnieM

## Overview

The ErnieM model was proposed in [ERNIE-M: Enhanced Multilingual Representation by Aligning
Cross-lingual Semantics with Monolingual Corpora](https://arxiv.org/abs/2012.15674)  by Xuan Ouyang, Shuohuan Wang, Chao Pang, Yu Sun,
Hao Tian, Hua Wu, Haifeng Wang.

The abstract from the paper is the following:

> *Recent studies have demonstrated that pre-
> trained cross-lingual models achieve impres-
> sive performance in downstream cross-lingual
> tasks. This improvement benefits from learn-
> ing a large amount of monolingual and par-
> allel corpora. Although it is generally ac-
> knowledged that parallel corpora are critical
> for improving the model performance, ex-
> isting methods are often constrained by the
> size of parallel corpora, especially for low-
> resource languages. In this paper, we pro-
> pose ERNIE-M, a new training method that
> encourages the model to align the representa-
> tion of multiple languages with monolingual
> corpora, to overcome the constraint that the
> parallel corpus size places on the model per-
> formance. Our key insight is to integrate
> back-translation into the pre-training process.
> We generate pseudo-parallel sentence pairs on
> a monolingual corpus to enable the learning
> of semantic alignments between different lan-
> guages, thereby enhancing the semantic mod-
> eling of cross-lingual models. Experimental
> results show that ERNIE-M outperforms ex-
> isting cross-lingual models and delivers new
> state-of-the-art results in various cross-lingual
> downstream tasks.*

Reviewer: ```suggestion
*Recent studies have demonstrated that pre-trained cross-lingual models achieve impressive performance in downstream cross-lingual tasks. 
This improvement benefits from learning a large amount of monolingual and par-
allel corpora. Although it is generally acknowledged that parallel corpora are critical
for improving the model performance, ex-
isting methods are often constrained by the
size of parallel corpora, especially for low-
resource languages. In this paper, we pro-
pose ERNIE-M, a new training method that
encourages the model to align the representa-
tion of multiple languages with monolingual
corpora, to overcome the constraint that the
parallel corpus size places on the model per-
formance. Our key insight is to integrate
back-translation into the pre-training process.
We generate pseudo-parallel sentence pairs on
a monolingual corpus to enable the learning
of semantic alignments between different lan-
guages, thereby enhancing the semantic mod-
eling of cross-lingual models. Experimental
results show that ERNIE-M outperforms ex-
isting cross-lingual models and delivers new
state-of-the-art results in various cross-lingual
downstream tasks.*
```
etc. the additional dashes should not be there ;)

*Recent studies have demonstrated that pre-trained cross-lingual models achieve impressive performance in downstream cross-lingual tasks. This improvement benefits from learning a large amount of monolingual and parallel corpora. Although it is generally acknowledged that parallel corpora are critical for improving the model performance, existing methods are often constrained by the size of parallel corpora, especially for lowresource languages. In this paper, we propose ERNIE-M, a new training method that encourages the model to align the representation of multiple languages with monolingual corpora, to overcome the constraint that the parallel corpus size places on the model performance. Our key insight is to integrate back-translation into the pre-training process. We generate pseudo-parallel sentence pairs on a monolingual corpus to enable the learning of semantic alignments between different languages, thereby enhancing the semantic modeling of cross-lingual models. Experimental results show that ERNIE-M outperforms existing cross-lingual models and delivers new state-of-the-art results in various cross-lingual downstream tasks.*

Tips:

> 1. Ernie-M is BERT-like model so it is stacked Transformer Encoder.

Reviewer 2: ```suggestion
1. Ernie-M is a BERT-like model so it is a stacked Transformer Encoder.
```

> 2. Instead of using MaskedLM for pretraining (like BERT) the authors used two novel techniques: `Cross-attention Masked Language Modeling` and `Back-translation Masked Language Modeling`

Author: I didn't find those LMHead's code on the [official paddlenlp implementation](https://github.com/PaddlePaddle/PaddleNLP/blob/2992a811c9aaefe7e1bc0502368e6c737d997a38/paddlenlp/transformers/ernie_m/modeling.py) but they were still used in pre-training. For now my goal is to make the model ready for fine-tuning, later if I find those implementations I will add them with new PRs . As you said I will modify the file as - 
``
2. Instead of using MaskedLM for pretraining (like BERT) the authors used two novel techniques: `Cross-attention Masked Language Modeling` and `Back-translation Masked Language Modeling (For now these two LMHead implementations are not available in `huggingface-transformers` but you can FineTune ErnieM!)
``

Is it ok?

> 2. Instead of using MaskedLM for pretraining (like BERT) the authors used two novel techniques: `Cross-attention Masked Language Modeling` and `Back-translation Masked Language Modeling`

Reviewer 3: Yes it's the best approach 😉

> 2. Instead of using MaskedLM for pretraining(like BERT) the authors used two novel techniques such as `Cross-attention Masked Language Modeling` and `Back-translation Masked Language Modeling`

Reviewer 2: ```suggestion
2. Instead of using MaskedLM for pretraining (like BERT) the authors used two novel techniques: `Cross-attention Masked Language Modeling` and `Back-translation Masked Language Modeling`
```

3. It is a multilingual language model.

> 4. NSP was not used in pretraining process.

Reviewer 3: The acronym should be defined

This model was contributed by [Susnato Dhar](https://huggingface.co/susnato). The original code can be found [here](https://github.com/PaddlePaddle/PaddleNLP/tree/develop/paddlenlp/transformers/ernie_m).

## ErnieMConfig

[[autodoc]] ErnieMConfig

## ErnieMTokenizer

[[autodoc]] ErnieMTokenizer
    - build_inputs_with_special_tokens
    - get_special_tokens_mask
    - create_token_type_ids_from_sequences
    - save_vocabulary

## ErnieMModel

[[autodoc]] ErnieMModel
    - forward

## ErnieMForSequenceClassification

> [[autodoc]] transformers.ErnieMForSequenceClassification
>     - forward
> 
> 
> ## ErnieMForMultipleChoice
> 
> [[autodoc]] transformers.ErnieMForMultipleChoice
>     - forward
> 
> 
> ## ErnieMForTokenClassification
> 
> [[autodoc]] transformers.ErnieMForTokenClassification
>     - forward

Reviewer 2: ```suggestion
[[autodoc]] ErnieMForSequenceClassification
    - forward

## ErnieMForMultipleChoice

[[autodoc]] ErnieMForMultipleChoice
    - forward

## ErnieMForTokenClassification

[[autodoc]] ErnieMForTokenClassification
    - forward
```
